### PR TITLE
Check pattern refutability on THIR

### PR DESCRIPTION
**Pull request by cjgillot**
The current `check_match` query is based on HIR, but partially re-lowers HIR into THIR.
This PR proposed to use the results of the `thir_body` query to check matches, instead of re-building THIR.

Most of the diagnostic changes are spans getting shorter, or commas/semicolons not getting removed.

This PR degrades the diagnostic for confusing constants in patterns (`let A = foo()` where `A` resolves to a `const A` somewhere): it does not point ot the definition of `const A` any more.

### CODE REVIEW

**Comment by cjgillot** [comment]
r? compiler

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 1]
```diff
@@ -390,58 +421,45 @@ impl<'p, 'tcx> MatchVisitor<'_, 'p, 'tcx> {
...
+                    variable: snippet,
+                });
+            }
+        }
+
+        if let Some(span) = sp
+            && self.tcx.sess.source_map().is_span_accessible(span)
+            && interpreted_as_const.is_none()
+        {
+            let mut bindings= vec![];
```
Comment (Nilstrieb) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:449]: nit:
```suggestion
            let mut bindings = vec![];
```

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 2]
```diff
@@ -390,58 +421,45 @@ impl<'p, 'tcx> MatchVisitor<'_, 'p, 'tcx> {
...
+            pat.each_binding(|name, _, _, _| bindings.push(name));
+
+            let semi_span = span.shrink_to_hi();
```
Comment (Nilstrieb) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:452]: Before, this actually replaced the semicolon with the if let body, now there's still a semicolon hanging around. It would be nice if the suggestion still had the semicolon removed.

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs [hunk 3]
```diff
@@ -877,35 +897,31 @@ fn is_binding_by_move(cx: &MatchVisitor<'_, '_, '_>, hir_id: HirId) -> bool {
...
-fn check_borrow_conflicts_in_at_patterns(cx: &MatchVisitor<'_, '_, '_>, pat: &Pat<'_>) {
+fn check_borrow_conflicts_in_at_patterns<'tcx>(cx: &MatchVisitor<'_, '_, 'tcx>, pat: &Pat<'tcx>) {
     // Extract `sub` in `binding @ sub`.
-    let (name, sub) = match &pat.kind {
-        hir::PatKind::Binding(.., name, Some(sub)) => (*name, sub),
-        _ => return,
-    };
-    let binding_span = pat.span.with_hi(name.span.hi());
+    let PatKind::Binding { name, mode, ty, subpattern: Some(box ref sub), .. } = pat.kind else { return };
+    let binding_span = pat.span; //.with_hi(name.span.hi());
```
Comment (Nilstrieb) [inline compiler/rustc_mir_build/src/thir/pattern/check_match.rs:903]: Why is this commented out here?

File: compiler/rustc_middle/src/dep_graph/mod.rs [hunk 4]
```diff
@@ -94,7 +94,7 @@ impl<'tcx> DepContext for TyCtxt<'tcx> {
     }
 
     #[inline]
-    fn dep_kind_info(&self, dep_kind: DepKind) -> &DepKindStruct<'tcx> {
-        &self.query_kinds[dep_kind as usize]
+    fn dep_kind_info(&self, dk: DepKind) -> &DepKindStruct<'tcx> {
+        &self.query_kinds[dk as usize]
```
Comment (Nilstrieb) [inline compiler/rustc_middle/src/dep_graph/mod.rs:98]: Why was this necessary?

**Review by Noratrieb — COMMENTED** [review]
That's some great cleanup, thanks! I especially love how some of the diagnostics got better, like how `let None = Some(());` now suggests `if let` when it didn't previously.

File: compiler/rustc_mir_build/src/build/mod.rs [hunk 5]
```diff
@@ -61,6 +61,7 @@ fn mir_build(tcx: TyCtxt<'_>, def: ty::WithOptConstParam<LocalDefId>) -> Body<'_
         ty::WithOptConstParam { did, const_param_did: None } => {
             tcx.ensure().thir_check_unsafety(did);
             drop(tcx.thir_abstract_const(did));
+            tcx.ensure().check_match(did);
```
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_mir_build/src/build/mod.rs:64]: `check_match` only fetches `thir_body(ty::WithOptConstParam::unknown(did))`, so we only need to ensure if we are stealing this exact one.

File: compiler/rustc_mir_build/src/thir/pattern/check_match.rs (hunk 2 above)
Comment (cjgillot) [reply compiler/rustc_mir_build/src/thir/pattern/check_match.rs:452]: THIR spans don't contain the semicolon any more. Removing it means trying to re-parse the source looking for it. OTOH, RustFmt should be able to do the cleanup on its own.

File: compiler/rustc_mir_build/src/build/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Nilstrieb) [reply compiler/rustc_mir_build/src/build/mod.rs:64]: Makes sense, const args can't contain a match. Is the loop over all body owners in `rustc_interface` still needed then? It seems redundant (except maybe for time passes niceness)

File: compiler/rustc_middle/src/dep_graph/mod.rs (hunk 4 above)
Comment (cjgillot) [reply compiler/rustc_middle/src/dep_graph/mod.rs:98]: There is a `DepKind::dep_kind` variant. The former version did not handle hygiene correctly, so did not fire in case of macro expansion. The new version just ignores hygiene, so fires. (In that case, it's actually the more correct behaviour.)

File: compiler/rustc_mir_build/src/build/mod.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cjgillot) [reply compiler/rustc_mir_build/src/build/mod.rs:64]: AnonConsts can definitely contain patterns to check. If we want to get rid of the `rustc_interface` loop, we'd need to force `check_match` in both cases, with a `WithOptConstParam` as argument. Like we do for `thir_check_unsafety`. I tried to avoid writing all the boilerplate just to remove it once https://github.com/rust-lang/rust/pull/96840 lands.

[the remark this answers is not in the document]
Comment (Nilstrieb) [reply compiler/rustc_mir_build/src/build/mod.rs:64]: Ah right, makes sense. Seems fine to keep it as-is then

**Comment by michaelwoerister** [comment]
Same here.
r? compiler

**Comment by michaelwoerister** [comment]
You should give that another try, rustbot 😉
r? compiler

**Comment by eholk** [comment]
This is probably outside my ability to review well...

@compiler-errors - is this something you could handle?

r? @compiler-errors

**Review by compiler-errors — APPROVED** [review]

**Comment by compiler-errors** [comment]
Everything here looks fine. I appreciate the clarity of the code afterwards, the primary regressions are just helloworld, and the `WithOptConstParam` treatment in this PR looks fine. r=me after @Nilstrieb said he wanted to give this one final pass over, or if he doesn't get to it by the end of the week :)

r? @Nilstrieb

**Review by Noratrieb — APPROVED** [review]

**Review by Noratrieb — APPROVED** [review]

**Comment by rylev** [comment]
Marking as triaged since the improvements outweigh the regressions and the regressions are small and in secondary benchmarks.

@rustbot label: +perf-regression-triaged
